### DISCUSSION TITLE
why are some libs acquired differently to others?

### DISCUSSION THREAD

**Discussion by kjm1102** [micropython/micropython#13047] [Libraries & Drivers]
Sometimes getting a lib in upython is relatively painless, eg
```
import mip
mip.install('github:micropython/micropython-esp32-ulp')
```
but other libs like ulab can only be acquired by finding a version of upython with it already baked in or trying to compile your own version from scratch. 

It occurs to me that I don't understand why some libs, like ulab, can't be mipped?

**Comment by andrewleech** [comment]
Packages like `ulab` and `lvgl` are C libraries that have to be compiled against / into micropython, wheras the packages in `micropython-lib` for instance are all pure-python so can be turned into bytecode mpy files.

There's the "in-between" packages that are C libraries written as dynamic c modules that does turn them into mpy files that can be copied onto a unit, but these need to be compiled separatly for each architecture and the correct one used. 
However the current deployment system with micropython-lib and mip doesn't have full support for these yet but hopefully will in future. 
Also, not all C libraries can be easily built this way, it does have limitations and uses more ram / flash on the unit than the more common "compiled in" approach to C libraries.

**Reply by kjm1102** [reply]
Tnx Andrew. The C+ crowd have a nice .h lib for motion detection. Am I right in thinking it could be built into a custom micropython or would it not have the hooks to access it from python.

**Reply by andrewleech** [reply]
It's pretty safe to say won't have hooks / python functions defined to access the C functions, but these can often be added to act as python wrapper functions to access C ones as appropriate.

https://docs.micropython.org/en/latest/develop/cmodules.html
https://github.com/micropython/micropython/tree/master/examples/usercmodule

**Reply by mattytrentini** [reply]
> The C+ crowd have a nice .h lib for motion detection. 

I'm curious! Which .h lib are you referring to?

**Reply by kjm1102** [reply]
https://github.com/eloquentarduino/EloquentSurveillance

**Reply by andrewleech** [reply]
That does look like a neat project, the motion code is nicely self-contained https://github.com/eloquentarduino/EloquentSurveillance/blob/main/src/Motion.h

It's all c++ so not as trivially easy to pull into a micropython c module, but could be wrapped and exposed following some of the micropython/examples/usercmodule/cppexample guidance. That example doesn't include how to use and expose a class as C.... you might even find it easier to squash that motion class down into flat C functions instead and remove the c++ usage.

Note it's GPL so anything derived from it can't be submitted as a library in micropython-lib but it's certainly fine for any personal / published open source projects - just be sure to give appropriate attribution!

**Reply by kjm1102** [reply]
A bit of a stretch for my skill level Andrew. Might need to find a better pir sensor & abandon my dreams of comparing jpgs for differences to validate the pir wakeup.

I must say I'm a bit envious of what the C+ crowd are able to achieve on esp32 though  https://github.com/0015/ESP32-OpenCV-Projects

**Reply by andrewleech** [reply]
Having that show direct on screen is a neat effect, though watching their actual video on youtube shows the framerate isn't quite as smooth as the short video direct on the github page 😛

You should take a look at OpenMV - it does this kind of on-device camera processing with micropython!
It's also got an amazing live-streaming IDE to make it really easy to use.

https://www.youtube.com/watch?v=2xOxaPfdg2Q
https://www.youtube.com/watch?v=hdlU92c6ny0
https://www.youtube.com/watch?v=h9VoDcROkxc

I haven't seen many people driving direct to screen, but some support is built in too, like here:
https://www.youtube.com/watch?v=21gAfCXcHSk

It can do some neat machine learning stuff too.
https://www.youtube.com/watch?v=YKwVope5RsU
